Dockerfile

```diff
@@ -1,78 +1,137 @@
+    echo "Using V8 download architecture: ${MAPPED_ARCH} based on TARGETARCH: ${TARGETARCH}" && \
+    # Download V8 using mapped arch
+    curl --fail -L -o /libc_v8.a https://github.com/lightpanda-io/zig-v8-fork/releases/download/${ZIG_V8}/libc_v8_${V8}_linux_${MAPPED_ARCH}.a
+
+# Stage 3: Download Zig and Minisign
+FROM base AS zig_download
+
+ARG MINISIG
+ARG ZIG
+ARG ZIG_MINISIG
+ARG TARGETARCH # Declare TARGETARCH to use it in this stage
+
+# Install dependencies needed for Zig and Minisign install
+RUN apt-get update -yq && \
+    apt-get install -yq --no-install-recommends curl xz-utils ca-certificates
+
+# Install Minisign and Zig
+RUN case ${TARGETARCH} in \
+      amd64) MAPPED_ARCH=x86_64 ;; \
+      arm64) MAPPED_ARCH=aarch64 ;; \
+      *) MAPPED_ARCH=${TARGETARCH} ;; \
+    esac && \
+    echo "Using mapped architecture: ${MAPPED_ARCH} based on TARGETARCH: ${TARGETARCH}" && \
+    # install minisig
+    curl --fail -L -O https://github.com/jedisct1/minisign/releases/download/${MINISIG}/minisign-${MINISIG}-linux.tar.gz && \
+    tar xvzf minisign-${MINISIG}-linux.tar.gz && \
+    mv minisign-linux/${MAPPED_ARCH}/minisign /usr/local/bin/minisign && \
+    rm -fr minisign-${MINISIG}-linux.tar.gz minisign-linux && \
+    # install zig using mapped arch (MAPPED_ARCH)
+    curl --fail -L -O https://ziglang.org/download/${ZIG}/zig-linux-${MAPPED_ARCH}-${ZIG}.tar.xz && \
+    curl --fail -L -O https://ziglang.org/download/${ZIG}/zig-linux-${MAPPED_ARCH}-${ZIG}.tar.xz.minisig && \
+    minisign -Vm zig-linux-${MAPPED_ARCH}-${ZIG}.tar.xz -P ${ZIG_MINISIG} && \
+    tar xvf zig-linux-${MAPPED_ARCH}-${ZIG}.tar.xz && \
+    mv zig-linux-${MAPPED_ARCH}-${ZIG} /zig
+
+# Stage 4: Build stage
+FROM base AS builder
+
+ARG TARGETARCH # Declare TARGETARCH to use it in this stage
+
+# Install build dependencies (removed xz-utils, kept curl for potential other uses)
+RUN apt-get update -yq && \
+    apt-get install -yq \
+        python3 ca-certificates git \
+        pkg-config libglib2.0-dev \
+        gperf libexpat1-dev \
+        cmake clang \
+        curl git
 
-RUN make install-libiconv && \
-    make install-netsurf && \
-    make install-mimalloc
+# Copy installed Zig from the zig_download stage
+COPY --from=zig_download /zig /usr/local/lib/zig
+# Create the symlink in the builder stage
+RUN ln -s /usr/local/lib/zig/zig /usr/local/bin/zig
 
-# download and install v8
-RUN curl --fail -L -o libc_v8.a https://github.com/lightpanda-io/zig-v8-fork/releases/download/${ZIG_V8}/libc_v8_${V8}_linux_${ARCH}.a && \
-    mkdir -p v8/build/${ARCH}-linux/release/ninja/obj/zig/ && \
-    mv libc_v8.a v8/build/${ARCH}-linux/release/ninja/obj/zig/libc_v8.a
+# Copy source code
+COPY --from=source /browser /browser
+COPY --from=source /root/.gitconfig /root/.gitconfig
 
-# build release
-RUN make build
+WORKDIR /browser
 
-FROM ubuntu:24.04
+# Install build dependencies from source
+RUN make install-libiconv
+RUN make install-netsurf
+RUN make install-mimalloc
```

Comment:
Separated the `make` command because we can use build cache in each command.
We use multi-stage build so we don't have to worry about the final image size